src/Workspaces/Core/Portable/Rename/Renamer.SyncNamespaceDocumentAction.cs

```diff
@@ -0,0 +1,118 @@
+            {
+                var solution = document.Project.Solution;
+
+                if (!_analysis.SupportsSyncNamespace)
+                {
+                    return solution;
+                }
+
+                var changeNamespaceService = document.GetRequiredLanguageService<IChangeNamespaceService>();
+
+                var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+                var targetNamespace = BuildNamespaceFromFolders(document.Folders, syntaxFacts);
+
+                var syntaxRoot = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+                // Don't descend into anything other than top level declarations from the root.
+                // ChangeNamespaceService only controls top level declarations right now
+                var namespaceDeclarations = syntaxRoot.DescendantNodes(n => !syntaxFacts.IsDeclaration(n)).Where(n => syntaxFacts.IsNamespaceDeclaration(n));
+
+                foreach (var namespaceDeclaration in namespaceDeclarations)
+                {
+                    solution = await changeNamespaceService.ChangeNamespaceAsync(
+                        document,
+                        namespaceDeclaration,
+                        targetNamespace,
+                        cancellationToken).ConfigureAwait(false);
+
+                    document = solution.GetRequiredDocument(document.Id);
+                }
+
+                return solution;
+
+                static string BuildNamespaceFromFolders(
+                    IEnumerable<string> folders,
+                    ISyntaxFacts syntaxFacts)
+                {
+                    var parts = folders.SelectMany(folder => folder.Split(new[] { '.' }));
+
+                    // string.Empty is used to represent the global namespace
+                    // when calling ChangeNamespaceAsync
+                    return parts.All(syntaxFacts.IsValidIdentifier) ? string.Join(".", parts) : string.Empty;
+                }
+            }
+
+            public static Task<SyncNamespaceDocumentAction?> TryCreateAsync(Document document, IReadOnlyList<string> newFolders, OptionSet optionSet, CancellationToken _)
+            {
+                var errors = new ArrayBuilder<ErrorResource>();
+
+                var analysisResult = Analyze(document);
+
+                if (!analysisResult.SupportsSyncNamespace)
+                {
+                    return Task.FromResult<SyncNamespaceDocumentAction?>(null);
+                }
+
+                return Task.FromResult<SyncNamespaceDocumentAction?>(new SyncNamespaceDocumentAction(analysisResult, optionSet, errors.ToImmutableAndFree()));
+            }
+
+            private static AnalysisResult Analyze(Document document)
+            {
+                return new AnalysisResult(
+                    supportsSyncNamespace: document.Project.Language == LanguageNames.CSharp);
```

Comment:
1. why only C#?
2. this should front load the actual checks that the name change can happen.

Reply:
There's no sync namespace implementation for VB. There's an issue filed and noted in the tests